lake-primitives/src/types/block.rs

```diff
@@ -118,128 +92,124 @@ impl Block {
     /// the action chain has begun. Other indexer developers care about it because of the habits
     /// from other blockchains like Ethereum where a transaction is a main asset. In case of NEAR
     /// [Receipts](crate::receipts::Receipt) are more important.
-    pub fn transactions(&mut self) -> impl Iterator<Item = &transactions::Transaction> {
-        if self.transactions.is_empty() {
-            self.transactions = self
-                .streamer_message
-                .shards
-                .iter()
-                .filter_map(|shard| shard.chunk.as_ref().map(|chunk| chunk.transactions.iter()))
-                .flatten()
-                .map(TryInto::try_into)
-                .filter_map(|transactions| transactions.ok())
-                .collect();
-        }
+    pub fn transactions(&self) -> impl Iterator<Item = &transactions::Transaction> {
         self.transactions.iter()
     }
 
-    /// Internal method to build the cache of actions on demand
-    fn actions_from_streamer_message(&self) -> Vec<actions::Action> {
-        self.streamer_message()
-            .shards
-            .iter()
-            .flat_map(|shard| shard.receipt_execution_outcomes.iter())
-            .filter_map(|receipt_execution_outcome| {
-                actions::Action::try_vec_from_receipt_view(&receipt_execution_outcome.receipt).ok()
-            })
-            .flatten()
-            .collect()
-    }
-
     /// Returns an iterator of the [Actions](crate::actions::Action) executed in the [Block]
-    pub fn actions(&mut self) -> impl Iterator<Item = &actions::Action> {
-        if self.actions.is_empty() {
-            self.build_actions_cache();
-        }
+    pub fn actions(&self) -> impl Iterator<Item = &actions::Action> {
         self.actions.iter()
     }
 
-    /// Returns an iterator of the [Events](crate::events::Event) emitted in the [Block]
-    pub fn events(&mut self) -> impl Iterator<Item = &events::Event> {
-        if self.events.is_empty() {
-            self.build_events_hashmap();
-        }
-        self.events.values().flatten()
+    /// Returns a Vec of [Events](crate::events::Event) emitted in the [Block]
+    pub fn events(&self) -> HashMap<super::ReceiptId, Vec<events::Event>> {
```

Comment:
The doc and the return value are out of sync. I would keep the original return value (iterator)